### ISSUE TITLE
Kafka/H2 Port Conflict

### ISSUE THREAD

**Issue by ruddell** [jhipster/generator-jhipster#4208]
##### **Overview of the issue**
Starting an app that uses H2 database and then trying to start Kafka via the docker image results in a port conflict.  H2's default port is 9092 and Kafka's docker yml also uses 9092. 

If you start Kafka first, H2 uses a different port automatically and everything works fine.

##### **Motivation for or Use Case**

In dev mode for a microservice architecture, it is common to have at least one app running h2.

##### **Reproduce the error**

Launch the jhipster-registry and any app using H2
Run `docker-compose -f src/main/docker/kafka.yml up -d` in the app using Kafka 
Error stating port 9092 is already allocated: `Bind for 0.0.0.0:9092 failed: port is already allocated`

##### **Suggest a Fix**

Use a different port for H2 or Kafka, or add it to the documentation that you need to start Kafka first

##### **JHipster Version(s)**

v3.8.0 and master (kafka was new in v3.8.0)

##### **JHipster configuration, a `.yo-rc.json` file generated in the root folder**

Gateway (uses H2):
```
{
  "generator-jhipster": {
    "jhipsterVersion": "3.8.0",
    "baseName": "gway",
    "packageName": "com.mycompany.myapp",
    "packageFolder": "com/mycompany/myapp",
    "serverPort": "8080",
    "authenticationType": "jwt",
    "hibernateCache": "hazelcast",
    "clusteredHttpSession": false,
    "websocket": false,
    "databaseType": "sql",
    "devDatabaseType": "h2Disk",
    "prodDatabaseType": "mysql",
    "searchEngine": false,
    "messageBroker": false,
    "buildTool": "maven",
    "enableSocialSignIn": false,
    "jwtSecretKey": "c82be367131184467150c2fc56c5ad10e811f2d2",
    "useSass": false,
    "applicationType": "gateway",
    "testFrameworks": [
      "gatling"
    ],
    "jhiPrefix": "jhi",
    "enableTranslation": true,
    "nativeLanguage": "en",
    "languages": [
      "en"
    ]
  }
}
```

Microservice with Kafka enabled
```
{
  "generator-jhipster": {
    "jhipsterVersion": "3.8.0",
    "baseName": "kafkatest",
    "packageName": "com.mycompany.myapp",
    "packageFolder": "com/mycompany/myapp",
    "serverPort": "8008",
    "authenticationType": "jwt",
    "hibernateCache": "hazelcast",
    "clusteredHttpSession": false,
    "websocket": false,
    "databaseType": "sql",
    "devDatabaseType": "h2Disk",
    "prodDatabaseType": "mysql",
    "searchEngine": false,
    "messageBroker": "kafka",
    "buildTool": "maven",
    "enableSocialSignIn": false,
    "jwtSecretKey": "e61e9bb3401315c7b5ef6621da06e5cb19f54916",
    "enableTranslation": false,
    "applicationType": "microservice",
    "testFrameworks": [],
    "jhiPrefix": "jhi",
    "skipClient": true,
    "skipUserManagement": true
  }
}
```

##### **Entity configuration(s) `entityName.json` files generated in the `.jhipster` directory**

No entities required

##### **Browsers and Operating System**

macOS Sierra v10.12

**Comment by jdubois** [comment]
That's very annoying!!! Can't believe the guys coding Kafka didn't have a look at choosing a port that isn't already used by something as common as H2!!!!
For me H2 is much more important, and this is clearly Kafka's fault -> let's change the default Kafka port, and document it everywhere...

**Comment by jdubois** [comment]
This adds quite a lot of complexity, for #4206 for example.

I'm not sure anymore about this: OK the port conflict is real, but then I don't think we use the tcp server from H2. I thought it was done for "mvn liquibase:diff" but I couldn't find the configuration.

So there is still work to do here -> I'm reopening this

**Comment by jdubois** [comment]
I just can't change the Kafka port, with the current image we use...

- If I map to port 9093 on the host and advertise port 9093, then I have LEADER_NOT_AVAILABLE errors from Kafka when I post messages (as internally it uses 9092, so the advertised port is false)
- If I map to port 9093 and advertise port 9092, of course my clients can't see the good port and fail

Anyway that's always annoying not to use the default port, this will confuse everyone.

So I tried again with H2 and port 9092:

- I can't find an issue when running a JHipster app
- I tried running the JHipster Registry before Kafka, like is written here, but I had no issue
- The only issue I had was when running Sonar, which also uses port 9092 as it's also using H2. But then, it's strange to have Sonar and Kafka running at the same time.

So my current opinion is to switch back to port 9092 and just ignore this issue -> @ruddell can you comment on this? Are you sure you get the error with the registry?

**Comment by cbornet** [comment]
We could also change the H2 TCP port ?

**Comment by jdubois** [comment]
You are right... Then I'm not sure we configure the h2 tcp server at all...

Le 5 oct. 2016 6:50 PM, "Christophe Bornet" <notifications@github.com> a
écrit :

**Comment by ruddell** [comment]
This issue happens when you run an h2 app before starting kafka.yml.  You can start any type of app that uses h2 (registry does not have a database) and check that it uses port 9092 by running `lsof -i :9092`.

The issue only happens in dev and only when someone starts a non-Kafka h2 app before starting Kafka, so I think we should go back to the defaults.

**Comment by jdubois** [comment]
Oh OK then I understand.

I'll try to change the default H2 port to 9093 -> I don't think anybody uses it at the moment. In the future, that might be useful, as I wanted to use it for doing `./mvnw liquibase:diff` with H2 also (never had the time to finish this...).

Otherwise that's not a big deal, and for Kafka I have no other choice but use 9092 everywhere...

**Comment by ruddell** [comment]
@jdubois You can change the h2 TCP port by changing [line 83 in DatabaseConfiguration.java](https://github.com/jhipster/generator-jhipster/blob/393f0be79c53110f077eb90facd81a43a1702e1d/generators/server/templates/src/main/java/package/config/_DatabaseConfiguration.java#L83) to
`        return Server.createTcpServer("-tcp","-tcpAllowOthers", "-tcpPort", "9093");`

I can do a PR reverting this and setting the port if you'd like.

**Comment by jdubois** [comment]
Thanks @ruddell ! Yes if you have the time -> it's late in the evening here in France, I'll have a look at this tomorrow

**Comment by ruddell** [comment]
There can only be one H2 app running on 9093, so setting the H2 port causes another issue.  I think we should ignore it or note "start Kafka before any apps" in the docs.

[H2 tutorial](http://www.h2database.com/html/tutorial.html) says "If the default port [9092] is already in use, a random port is used."  That does not apply if you set the port.  Two apps trying to run with port 9093 throws a `java.net.BindException: Address already in use`.

**Comment by pascalgrimaud** [comment]
So much conflict!
I think we should keep the default port and just document it

**Comment by jdubois** [comment]
Yes OK. Anyway, you should start Kafka first, or your app won't start.

**Comment by jdubois** [comment]
This is in fact not finished, if you do microservices this causes another issue:

- I start microservice A in dev mode -> it uses port 9092
- I then start microservice B that uses Kafka -> port conflict!!!

This is so badly done, I don't understand why Kafka uses a port that is the same as something so common as H2!!
